src/devices/bus/nes/vrc_clones.cpp

```diff
@@ -548,6 +578,57 @@ void nes_th21311_device::write_h(offs_t offset, u8 data)
 		nes_konami_vrc2_device::write_h(offset, data);
 }
 
+/*-------------------------------------------------
+
+ Waixing San Guo Zhi Board, Henggedianzi UNL-SHJY3
+
+ Games: San Guo Zhi (mapper 252), several Chinese
+ hacks of Dragon Ball games (mapper 253)
+
+ VRC4 clones that substitute 2K of CHRRAM in place for
+ two of the standard 1K CHRROM pages. Unusually this is
+ done by a GAL looking at certain address bits of the
+ PPU's address space, as the PPU writes pattern tables.
+
+ iNES: mapper 252 and mapper 253
+
+ In MAME: Supported.
+
+ TODO: Why does Sanguozhi flicker every second or two?
+
+ -------------------------------------------------*/
+
+void nes_waixing_sgz_device::write_h(offs_t offset, u8 data)
+{
+	LOG_MMC(("waixing_sgz write_h, offset: %04x, data: %02x\n", offset, data));
+
+	nes_konami_vrc4_device::write_h(offset, data);
+
+	for (int bank = 0; bank < 8; bank++)
+		if ((m_mmc_vrom_bank[bank] & m_chr_mask) == m_chr_match)
+			chr1_x(bank, m_mmc_vrom_bank[bank] & 1, CHRRAM);
+}
+
+void nes_waixing_sgz_device::chr_w(offs_t offset, u8 data)
+{
+	device_nes_cart_interface::chr_w(offset, data);
+
+	switch (m_mmc_vrom_bank[BIT(offset, 10, 3)])
+	{
+		case 0x88:
+			m_chr_mask = 0xfc;
+			m_chr_match = 0x4c;
+			break;
+		case 0xc2:
+			m_chr_mask = 0xfe;
+			m_chr_match = 0x7c;
+			break;
+		case 0xc8:
+			m_chr_mask = 0xfe;
+			m_chr_match = 0x04;
+			break;
+	}
```

Comment:
On this variant where `m_chr_mask` and `m_chr_match` are both programmable, should they be reset on soft reset, or does it always program them?

Reply:
I can't say for sure. Here's the very confusing RE thread for more info: https://forums.nesdev.org/viewtopic.php?t=19041.

I can tell you that some of the 5 games here don't seem to write any address/bank value that triggers the switch and so they must depend on the power on values. Others write once or twice at boot and don't seem to again, and so they depend on the switch. So neither should care AFAICT about reset (not that this means they aren't reset).

That being said, to me it sounds like they've probed a couple GALs on actual boards but otherwise got the remaining values from debugging roms. It makes it seem like there's a common chip, but I'm not sure if they all have the same GAL or not. I'm not sure why the power on values are different. I'm not sure if the values the GAL(s) respond to is here is a reduced subset simply because these are the values the known VRC clone games respond to (for a more complete list used by similar MMC3 clone boards see: https://wiki.nesdev.org/w/index.php?title=INES_Mapper_195).

Reply:
If it’s actually a GAL, it won’t have a built-in power-on reset, so something must be getting it into the desired initial state.  Is the GAL reset by the reset line on the cartridge interface, or does it have a power-on reset circuit in the cartridge?  (Or is it something other than a GAL?)

Reply:
Oh, the GAL only seems to be there to latch/translate.  The actual logic comes from whatever’s under the glob.

Reply:
Unfortunately, I can't give anything but pure conjecture. A major point of note is that the two boards pictured in the discussion thread are NOT the boards for these 5 games. Same company, clearly related hardware with similar, very unique behavior. However, nobody seems to even have PCB labels for these 5 games, so it all seems to be guess work based on what the RE-ed boards do and what values these other related ROMs respond to in debugging.

Reply:
What's under the glob in the first board, FS306, is some Konami VRC4 clone. Here's the reference page on the board FWIW: https://wiki.nesdev.org/w/index.php?title=NES_2.0_Mapper_544. I guess these 5 games should be one similar boards, but without the extra features.